File: firmware/multi_deck/icons.cpp
#include "icons.h"

#include <lvgl.h>

namespace icons {
namespace {

struct Entry {
  const char *name;
  const char *symbol;
};

// Every symbol LVGL builds into the Montserrat fonts, named by the rule in icons.h. Sorted so
// that adding one is obvious and so this reads as the reference list it doubles as — the agent
// mirrors it in ICON_NAMES, and protocol_test.py parses this table to check the two agree.
//
// LV_SYMBOL_DUMMY is omitted deliberately: it is LVGL's internal marker for "no glyph", not
// something a tile should ever ask for.
const Entry kIcons[] = {
    {"audio", LV_SYMBOL_AUDIO},
    {"backspace", LV_SYMBOL_BACKSPACE},
    {"bars", LV_SYMBOL_BARS},
    {"battery_1", LV_SYMBOL_BATTERY_1},
    {"battery_2", LV_SYMBOL_BATTERY_2},
    {"battery_3", LV_SYMBOL_BATTERY_3},
    {"battery_empty", LV_SYMBOL_BATTERY_EMPTY},
    {"battery_full", LV_SYMBOL_BATTERY_FULL},
    {"bell", LV_SYMBOL_BELL},
    {"bluetooth", LV_SYMBOL_BLUETOOTH},
    {"bullet", LV_SYMBOL_BULLET},
    {"call", LV_SYMBOL_CALL},
    {"charge", LV_SYMBOL_CHARGE},
    {"close", LV_SYMBOL_CLOSE},
    {"copy", LV_SYMBOL_COPY},
    {"cut", LV_SYMBOL_CUT},
    {"directory", LV_SYMBOL_DIRECTORY},
    {"down", LV_SYMBOL_DOWN},
    {"download", LV_SYMBOL_DOWNLOAD},
    {"drive", LV_SYMBOL_DRIVE},
    {"edit", LV_SYMBOL_EDIT},
    {"eject", LV_SYMBOL_EJECT},
    {"envelope", LV_SYMBOL_ENVELOPE},
    {"eye_close", LV_SYMBOL_EYE_CLOSE},
    {"eye_open", LV_SYMBOL_EYE_OPEN},
    {"file", LV_SYMBOL_FILE},
    {"gps", LV_SYMBOL_GPS},
    {"home", LV_SYMBOL_HOME},
    {"image", LV_SYMBOL_IMAGE},
    {"keyboard", LV_SYMBOL_KEYBOARD},
    {"left", LV_SYMBOL_LEFT},
    {"list", LV_SYMBOL_LIST},
    {"loop", LV_SYMBOL_LOOP},
    {"minus", LV_SYMBOL_MINUS},
    {"mute", LV_SYMBOL_MUTE},
    {"new_line", LV_SYMBOL_NEW_LINE},
    {"next", LV_SYMBOL_NEXT},
    {"ok", LV_SYMBOL_OK},
    {"paste", LV_SYMBOL_PASTE},
    {"pause", LV_SYMBOL_PAUSE},
    {"play", LV_SYMBOL_PLAY},
    {"plus", LV_SYMBOL_PLUS},
    {"power", LV_SYMBOL_POWER},
    {"prev", LV_SYMBOL_PREV},
    {"refresh", LV_SYMBOL_REFRESH},
    {"right", LV_SYMBOL_RIGHT},
    {"save", LV_SYMBOL_SAVE},
    {"sd_card", LV_SYMBOL_SD_CARD},
    {"settings", LV_SYMBOL_SETTINGS},
    {"shuffle", LV_SYMBOL_SHUFFLE},
    {"stop", LV_SYMBOL_STOP},
    {"tint", LV_SYMBOL_TINT},
    {"trash", LV_SYMBOL_TRASH},
    {"up", LV_SYMBOL_UP},
    {"upload", LV_SYMBOL_UPLOAD},
    {"usb", LV_SYMBOL_USB},
    {"video", LV_SYMBOL_VIDEO},
    {"volume_max", LV_SYMBOL_VOLUME_MAX},
    {"volume_mid", LV_SYMBOL_VOLUME_MID},
    {"warning", LV_SYMBOL_WARNING},
    {"wifi", LV_SYMBOL_WIFI},
};

}  // namespace

const char *symbol(const String &name) {
  if (name.isEmpty()) return nullptr;

  // Linear over ~60 entries, and only while building a page. Not worth a sorted lookup.
  for (const Entry &entry : kIcons) {
    if (name == entry.name) return entry.symbol;
  }
  return nullptr;
}

}  // namespace icons

File: agent/deckbuilder/icons.py
"""Stand-in glyphs for the deck's built-in icons.

The deck's icons are LVGL's built-in symbols, which are FontAwesome glyphs compiled into the
firmware. None of that exists on the PC: there is no FontAwesome in the repo, and the compiled
font is a C array of bitmaps rather than something Pillow can open. So the preview substitutes
Segoe MDL2 Assets, which ships with Windows and covers most of the same vocabulary.

These are approximations and the window says so. That is a smaller compromise than it sounds,
because an icon's shape is the same in every theme — what the preview is actually for is
whether a colour, an opacity and a radius work together, and a stand-in glyph of the right size
in the right place answers that as well as the real one would.

Two rules keep the approximation honest:

  - every name the firmware knows has an entry here, checked by a test against ICON_NAMES the
    same way config.py's list is checked against icons.cpp. The damage from a bad guess is
    then bounded at "wrong picture", never "blank tile", because a blank tile in a preview
    reads as a layout bug you do not have.
  - anything unmapped, and anything the font turns out not to draw, falls back to a visible
    placeholder rather than to nothing.
"""

from __future__ import annotations

from pathlib import Path

# Windows ships both; the second is the newer name for the same idea. Checked in order.
FONT_CANDIDATES = (
    Path(r"C:\Windows\Fonts\segmdl2.ttf"),
    Path(r"C:\Windows\Fonts\SegoeIcons.ttf"),
)

# LVGL symbol name -> Segoe MDL2 Assets codepoint.
#
# A few are deliberate reinterpretations rather than matches. `tint` is a droplet in
# FontAwesome, but on this deck it is the theme-cycling button, so a colour palette says more
# about what the tile does. `eject` has no counterpart in this font and borrows the download
# arrow, which is the same gesture pointing at the same bar.
PREVIEW_GLYPHS: dict[str, str] = {
    "audio": "\ue8d6",
    "backspace": "\ue750",
    "bars": "\ue700",
    # E850-E859 is Battery0-Battery9, a monotonic fill ramp; E85A starts the *charging* series
    # over from empty, which is why the obvious "one past nine" pick for full draws less than
    # battery_3 does. Spread the deck's five levels across the ten.
    "battery_empty": "\ue850",
    "battery_1": "\ue852",
    "battery_2": "\ue855",
    "battery_3": "\ue857",
    "battery_full": "\ue859",
    "bell": "\ue7ed",
    "bluetooth": "\ue702",
    "bullet": "\ue91f",
    "call": "\ue717",
    "charge": "\ue945",
    "close": "\ue711",
    "copy": "\ue8c8",
    "cut": "\ue8c6",
    "directory": "\uf12b",
    "down": "\ue70d",
    "download": "\ue896",
    "drive": "\ueda2",
    "edit": "\ue70f",
    "eject": "\ue896",
    "envelope": "\ue715",
    "eye_close": "\ued1a",
    "eye_open": "\ue7b3",
    "file": "\ue7c3",
    "gps": "\ue707",
    "home": "\ue80f",
    "image": "\ueb9f",
    "keyboard": "\ue765",
    "left": "\ue76b",
    "list": "\ue8fd",
    "loop": "\ue8ee",
    "minus": "\ue738",
    "mute": "\ue74f",
    "new_line": "\ue751",
    "next": "\ue893",
    "ok": "\ue8fb",
    "paste": "\ue77f",
    "pause": "\ue769",
    "play": "\ue768",
    "plus": "\ue710",
    "power": "\ue7e8",
    "prev": "\ue892",
    "refresh": "\ue72c",
    "right": "\ue76c",
    "save": "\ue74e",
    "sd_card": "\ue7f1",
    "settings": "\ue713",
    "shuffle": "\ue8b1",
    "stop": "\ue71a",
    "tint": "\ue790",
    "trash": "\ue74d",
    "up": "\ue70e",
    "upload": "\ue898",
    "usb": "\uecf0",
    "video": "\ue714",
    "volume_max": "\ue995",
    "volume_mid": "\ue994",
    "warning": "\ue7ba",
    "wifi": "\ue701",
}


def font_path() -> Path | None:
    """The first icon font present, or None on a machine that has neither."""
    for candidate in FONT_CANDIDATES:
        if candidate.exists():
            return candidate
    return None


def glyph_for(name: str) -> str | None:
    """The stand-in character for an LVGL symbol name, or None to draw the placeholder."""
    return PREVIEW_GLYPHS.get(name)
